Implement complete dask.array.tile function

bouweandela opened the pull request:
This implements a complete `dask.array.tile` function, based on this https://github.com/dask/dask/issues/2447#issuecomment-500614803.

For example, 
```python
>>> from dask import array as da
>>> a = da.asarray([1, 2, 3])
>>> da.tile(a, [2, 1]).compute()
array([[1, 2, 3],
       [1, 2, 3]])
```
now works as expected.

- [ ] Tests added / passed
- [ ] Passes `black dask` / `flake8 dask`

Closes https://github.com/dask/dask/issues/2447

mrocklin commented on dask/array/tests/test_creation.py:
```diff
@@ -591,13 +597,29 @@ def test_tile_neg_reps(shape, chunks, reps):
+@pytest.mark.parametrize("a", [da.asarray([[[]]]), da.asarray([[], []])])
+@pytest.mark.parametrize("reps", [2, (3, 2, 5)])
+def test_tile_empty_array(a, reps):
+    assert_eq(np.tile(a, reps), da.tile(a, reps))
```
```suggestion
    assert_eq(np.tile(a.compute(), reps), da.tile(a, reps))
```

mrocklin replied:
As of Numpy 1.17 `np.tilie` will just dispatch to `da.tile` if given a dask array as an input, so let's coerce to numpy here first.

bouweandela replied:
That's a cool feature I didn't know about! Changed now.

mrocklin commented on dask/array/tests/test_creation.py:
```diff
@@ -571,9 +571,15 @@ def test_repeat():
         assert all(concat(d.repeat(r).chunks))
 
 
+@pytest.mark.parametrize("a", [da.asarray([0, 1, 2]), [[1, 2], [3, 4]]])
```
I have a slight preference to avoid creating dask array objects at import time.  This is probably fine, but I suspect that some future bugs might interfere more easily here.  Is it easy for you to bring this into the function?  Maybe in this one case we drop the parametrization.

bouweandela replied:
Thanks, dropped the parametrization.

mrocklin commented on dask/array/tests/test_creation.py:
```diff
@@ -571,9 +571,15 @@ def test_repeat():
+@pytest.mark.parametrize("a", [da.asarray([0, 1, 2]), [[1, 2], [3, 4]]])
+@pytest.mark.parametrize("reps", [2, (2, 2), (1, 2), (2, 1), (2, 3, 4, 0)])
+def test_tile_basic(a, reps):
+    assert_eq(np.tile(a, reps), da.tile(a, reps))
```
As below, we'll need to make sure that the argument to `np.tile` is a numpy array.

bouweandela replied:
Changed now.

bouweandela commented:
Thank you for your helpful review @mrocklin! I addressed your comments. Unfortunately the CI tests are failing now, but this seems unrelated to this pull request.

mrocklin commented:
Thanks for making the changes.  This looks good to me.  Merging in.

I'll take a look at the CI failures.  It looks like the recent bokeh release may have broken something.
